ion-schema/src/nfa.rs

```diff
@@ -0,0 +1,258 @@
+    pub(crate) fn from_elements_to_events<T: IntoIterator<Item = Element>>(iter: T) -> Vec<Event> {
+        let mut events: Vec<Event> = iter
+            .into_iter()
+            .map(<Element as Into<Event>>::into)
+            .collect();
+        // add END_OF_STREAM at the end of events vec
+        events.push(Event::EndOfStream);
+        events
+    }
+}
+
+impl From<Element> for Event {
+    fn from(value: Element) -> Self {
+        Event::Value(value)
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct Transition {
+    destination: usize,      // represents destination state
+    to_optional_state: bool, // represents if destination state is an optional state or not
+    type_id: Option<TypeId>, // None is used for final state transition
+}
+
+/// Represents an NFA that will eb used by the ordered_elements` constraint in order to validate an Ion Value
+#[derive(Default, Debug, Clone, PartialEq)]
+pub struct Nfa {
+    pub(crate) states: Vec<State>, // represents states with state_id
+    pub(crate) transitions: HashMap<usize, HashSet<Transition>>, // represents transitions between states
+}
+
+/// This is a context which will be used while validating an Ion value for `ordered_elements` constraint using its NFA
+#[derive(Debug, Clone)]
+pub struct NfaRun {
+    pub(crate) visits: HashMap<usize, usize>, // represents a map of (stat_id, visit_count)
+    pub(crate) nfa: Rc<Nfa>,                  // reference to the NFA
+}
+
+impl NfaRun {
+    pub fn new(nfa: Rc<Nfa>) -> Self {
+        Self {
+            visits: {
+                let mut visits = HashMap::new();
+                visits.insert(0, 1);
+                visits
+            },
+            nfa,
+        }
+    }
+
+    pub fn has_final_state(&self) -> bool {
+        self.visits
+            .iter()
+            .any(|(id, _)| self.nfa.states[*id] == State::Final)
+    }
+
+    pub fn transition(&mut self, event: Event, type_store: &TypeStore) -> HashMap<usize, usize> {
+        let mut new_states = HashMap::new();
+        for (source_state_id, visits) in self.visits.iter() {
+            let can_exit = self.nfa.states[*source_state_id].allows_exit_after_n_visits(*visits);
+            let transitions: &HashSet<Transition> =
+                self.nfa.transitions.get(source_state_id).unwrap();
+            self.evaluate_transitions(
+                *source_state_id,
+                transitions,
+                can_exit,
+                *visits,
+                event.to_owned(),
+                type_store,
+                &mut new_states,
+            );
+        }
+        new_states
+    }
+
+    fn evaluate_transitions(
+        &self,
+        source_state_id: usize,
+        transitions: &HashSet<Transition>,
+        can_exit: bool,
+        visits: usize,
+        event: Event,
+        type_store: &TypeStore,
+        new_states: &mut HashMap<usize, usize>,
```

Comment:
Clippy says there are too many arguments. At minimum, I don't think we need to pass around the set of transitions or the type store.

The set of transitions is unnecessary since it can be derived using the source state id:
```
let transitions: &HashSet<Transition> = self.nfa.transitions.get(source_state_id).unwrap();
```

I think the typestore could be store by reference in the `Nfa` struct (unless there's a circular reference problem). I don't see how there would be a circular reference problem storing it in `NfaRun` though, so that could be another option.